addons/binding/org.openhab.binding.pentair/src/main/java/org/openhab/binding/pentair/handler/PentairIntelliFloHandler.java

```diff
@@ -0,0 +1,178 @@
+            logger.debug("Intellflo received refresh command");
+            updateChannel(channelUID.getId(), null);
+        }
+    }
+
+    @Override
+    public void processPacketFrom(PentairPacket p) {
+
+        switch (p.getAction()) {
+            case 1: // Pump command - A5 00 10 60 01 02 00 20
+                logger.trace("Pump command (ack): {}: ", p.toString());
+                break;
+
+            case 4: // Pump control panel on/off
+                logger.trace("Turn pump control panel (ack) {}: {} - {}", p.getSource(),
+                        p.getByte(PentairPacket.STARTOFDATA), p.toString());
+                break;
+            case 5: // Set pump mode
+                logger.trace("Set pump mode (ack) {}: {} - {}", p.getSource(), p.getByte(PentairPacket.STARTOFDATA),
+                        p.toString());
+                break;
+            case 6: // Set run mode
+                logger.trace("Set run mode (ack) {}: {} - {}", p.getSource(), p.getByte(PentairPacket.STARTOFDATA),
+                        p.toString());
+                break;
+            case 7: // Pump status (after a request)
+                if (p.getLength() != 15) {
+                    logger.debug("Expected length of 15: ", p.toString());
+                    return;
+                }
+
+                /*
+                 * P: A500 d=10 s=60 c=07 l=0f 0A0602024A08AC120000000A000F22 <028A>
+                 * RUN 0a Started
+                 * MOD 06 Feature 1
+                 * PMP 02 ? drive state
+                 * PWR 024a 586 WATT
+                 * RPM 08ac 2220 RPM
+                 * GPM 12 18 GPM
+                 * PPC 00 0 %
+                 * b09 00 ?
+                 * ERR 00 ok
+                 * b11 0a ?
+                 * TMR 00 0 MIN
+                 * CLK 0f22 15:34
+                 */
+
+                logger.debug("Pump status: {}", p.toString());
+
+                PentairPacket_PumpStatus ppsOld = ppscur;
+                ppscur = new PentairPacket_PumpStatus(p);
+
+                updateChannel(INTELLIFLO_RUN, ppsOld);
+                updateChannel(INTELLIFLO_MODE, ppsOld);
+                updateChannel(INTELLIFLO_DRIVESTATE, ppsOld);
+                updateChannel(INTELLIFLO_POWER, ppsOld);
+                updateChannel(INTELLIFLO_RPM, ppsOld);
+                updateChannel(INTELLIFLO_PPC, ppsOld);
+                updateChannel(INTELLIFLO_ERROR, ppsOld);
+                updateChannel(INTELLIFLO_TIMER, ppsOld);
+
+                break;
+            default:
+                logger.debug("Unhandled Intelliflo command: {}", p.toString());
+                break;
+        }
+    }
+
+    @SuppressWarnings("null")
+    public void updateChannel(String channel, PentairPacket p) {
+        PentairPacket_PumpStatus pps = (PentairPacket_PumpStatus) p;
+
+        switch (channel) {
+            case INTELLIFLO_RUN:
+                if (pps == null || (pps.run != ppscur.run)) {
```

Comment:
Can you clarify what is going on here and why, it looks like we are always a little bit behind with updating the values from the thread or am I missing some detail here?